p2p/server.py

```diff
@@ -0,0 +1,125 @@
+import asyncio
+import os
+import logging
+import sys
+
+from typing import Tuple
+from eth_keys import (
+    datatypes,
+    keys,
+)
+from eth_utils import (
+    decode_hex,
+    keccak,
+)
+
+from p2p.auth import (
+    HandshakeResponder,
+    decode_auth_plain,
+    decode_auth_eip8,
+)
+from p2p.cancel_token import CancelToken
+from p2p.constants import (
+    ENCRYPTED_AUTH_MSG_LEN,
+    HASH_LEN,
+)
+from p2p.ecies import ecdh_agree
+from p2p.kademlia import (
+    Address,
+    Node,
+)
+from p2p.utils import sxor
+
+
+class Server:
+    """Server listening for incoming connections"""
+    logger = logging.getLogger("p2p.server.Server")
+    _server_address = ('localhost', 10000)
+    _connect_loop_sleep = 2
+    
+    def __init__(self, privkey: datatypes.PrivateKey) -> None:
+        self.privkey = privkey
+        self.cancel_token = CancelToken('Server')
+        self.incoming_connections = []
+
+    async def stop(self) -> None:
+        self.logger.info("Closing server...")
+        self.cancel_token.trigger()
+        await asyncio.sleep(1)
+
+    async def run(self) -> None:
+        self.logger.info("Running server...")
+        loop = asyncio.get_event_loop()
+        factory = asyncio.start_server(self.receive_handshake, *self._server_address)
+        asyncio.ensure_future(factory)
+        
+        while not self.cancel_token.triggered:
+            try:
+                loop.run_forever()
+            except:
+                self.logger.error("Unexpected error.")
+            await awyncio.wait([self.cancel_token.wait()], timeout=_connect_loop_sleep)
+
+    async def receive_handshake(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
+        # Use reader to read the auth_init msg !!!CONSTANT???!!!
+        msg = await reader.read(437)
+
+        # Use HandshakeResponder.decode_authentication(auth_init_message) on auth init msg
+        ephem_pubkey, initiator_nonce, initiator_pubkey = _decode_authentication(msg, self.privkey)
+
+        # Get remote's address 
+        ip, udp, _, _ = writer.get_extra_info("peername")
+        remote_address = Address(ip, udp)
+
+        # Create a `HandshakeResponder(remote: kademlia.Node, privkey: datatypes.PrivateKey)` instance
+        initiator_remote = Node(initiator_pubkey, remote_address)
+        responder = HandshakeResponder(initiator_remote, self.privkey)
+
+        # Call `HandshakeResponder.create_auth_ack_message(nonce: bytes)` to create the reply
+        responder_nonce = keccak(os.urandom(HASH_LEN))
```

Comment:
we recently realized we can use the new [`secrets.token_bytes`](https://docs.python.org/3/library/secrets.html#secrets.token_bytes) functions for things like this.